### ISSUE TITLE
Improved handling of `_FillValue` ?

### ISSUE THREAD

**Issue by ChrisBarker-NOAA** [Unidata/netcdf4-python#1374]
As it stands, if you create a variable without setting a fill_value (fill_value=None), then the variable will be created using the default fill value, but not create a `_FillValue` attribute.

This is a shame, as when reading the file, it's kind hard to know what the _FillValue is -- at least with netCDF4 -- you'd need to look it up in the default_fillvals dict. And that's a bit hard to do, as the dict in keyed wby the dtype code, whereas a_variable.dtype returns the dtype object:

```
In [42]: nc4.default_fillvals
Out[42]: 
{'S1': '\x00',
 'i1': -127,
 'u1': 255,
 'i2': -32767,
 'u2': 65535,
 'i4': -2147483647,
 'u4': 4294967295,
 'i8': -9223372036854775806,
 'u8': 18446744073709551614,
 'f4': 9.969209968386869e+36,
 'f8': 9.969209968386869e+36}

In [43]: bob.dtype
Out[43]: dtype('float32')
```
in fact, the easiest way I could determine the _FillValue was this:
```
In [52]: dtype = bob.dtype

In [53]: fill_value = nc4.default_fillvals[f"{dtype.char}{dtype.itemsize}"]

In [54]: fill_value
Out[54]: 9.969209968386869e+36
```

not very straightforward.

Anyway, all of this is a bit awkward, so some ideas (maybe some of this belongs in the C lib?):

1) Have createVariable always create a _FillValue attribute -- you can't set it later, but you can delete later if you really want.

2) if not (1), then have a way to specify -- "use the default fill value, and set the attribute" perhaps: `fill_value='default` (or `True`?)

3) have an easier way to get the fill value if the attribute isn't set:
   Variable.get_fill_value(), perhaps ?

If you all are open to any of these ideas, I could work on a PR.

NOTE: (3) would still be nice, even if 1 or 2 is done, as there are who knows how many files out in the wild, and will be for a long time, that don't have a _FillValue attribute.

**Comment by jswhit** [comment]
From the netcdf-c [docs](https://docs.unidata.ucar.edu/netcdf-c/current/attribute_conventions.html)  "It is not necessary to define your own _FillValue attribute for a variable if the default fill value for the type of the variable is adequate".  Always creating a ``_FillValue`` attribute, even when the user does not specify one, would be easy to do but a departure from the way the C library works.   I would prefer your option #3.

**Comment by ChrisBarker-NOAA** [comment]
Well, option (3) is a good idea regardless -- no matter how you slice it, there are a lot of files out there now that don't have a _FillValue set -- and it would be nice if there was an easy way to get the fill value in any case.

Should I try to write a PR? I do confess that I do not know how to use the C API for this.

 "It is not necessary to define your own _FillValue attribute for a variable if the default fill value for the type of the variable is adequate"

The problem there is that it's not necessary to define the fill value itself, but it's a bad idea to not have the attribute set. I think this is a bit of a relic of the early days:

The fill value is just what it implies -- the value used to fill the array before any values are set. So when writing a file, "just use the default' is a perfectly reasonable thing to do -- and why set _FillValue

However, these days, the `_FillValue` attribute  is often used, by clients reading the code, to signify "missing values" in which case, it's critical to know what it is -- and if there is no _FillValue attribute, it's pretty awkward to do so -- (3) would help, but it's still pretty darn clunky to write client code that will work with most files.

So I think it's a best practice to specify the `_FillValue` attribute, even if you are using the default, and it would be good to make that easy, hence (2).

As for (1) -- I still think that's a good idea, but as you say, it's not what the C lib does, so probably best to not do that -- though I may put in a Issue with the C lib about it :-)

**Comment by jswhit** [comment]
PR #1375 (as yet untested) implements a get_fill_value Variable method (option 3).  We both agree option 1 is probably not a good idea.  I'm not sold on option 2 either - we would essentially be over-riding the C library.  What if the default fill values change in the C lib?

**Comment by ChrisBarker-NOAA** [comment]
Thanks for the PR -- I added a couple comments.

> (1) would be overriding the C lib

 -- well, not overriding, but providing a different API.

I don't think (2) is doing so -- I don't think there's anything in the C lib (or the NUG) that says you should not set `_FillValue` if you are using the default. And I'm pretty sure you can set the `_FillValue` directly after creating the array, as indicated in the discussion of this closed issue:

https://github.com/Unidata/netcdf-c/issues/3038

"It is not necessary to define your own `_FillValue` attribute for a variable if the default fill value for the type of the variable is adequate. "

To me says that you don't *have* to do it, not that it's a bad idea *to* do it.

Which brings me back to (2) -- provide a way for the user to do it (use the default but also set the attribute), if they want to, which is very hard to do as it stands.

Thanks,
-CHB

**Comment by dopplershift** [comment]
As noted in the [netCDF docs](https://docs.unidata.ucar.edu/netcdf-c/current/attribute_conventions.html https://docs.unidata.ucar.edu/netcdf-c/current/attribute_conventions.html):

> Attribute names commencing with underscore ('_') are reserved for use by the netCDF library.

So seeing as how netcdf4-python isn't an independent library implementation of netCDF but relies on the C library, let's leave any writing of `_FillValue` to the C library.

I think it's perfectly fine to raise that [there](https://github.com/Unidata/netcdf-c) though.

**Comment by ChrisBarker-NOAA** [comment]
@dopplershift: overall, the C lib is lower level, you can set the _FillValue yourself however you want. This is just making it easy to do a common action that one would have to do by hand in C.

That being said, I may put a request in the C lib as well, for something similar.

**Comment by jswhit** [comment]
closed by pr #1375
